firedrake/linear_solver.py

```diff
@@ -101,27 +100,36 @@ def __init__(self, A, P=None, solver_parameters=None,
         self.set_from_options(self.ksp)
 
     @cached_property
-    def _b(self):
-        """A function to store the RHS.
-
-        Used in presence of BCs."""
-        return function.Function(self._W)
+    def test_space(self):
+        return self.A.a.arguments()[0].function_space()
 
     @cached_property
-    def _Abcs(self):
-        """A function storing the action of the operator on a zero Function
-        satisfying the BCs.
+    def trial_space(self):
+        return self.A.a.arguments()[1].function_space()
 
-        Used in the presence of BCs.
-        """
-        b = function.Function(self._W)
-        for bc in self.A.bcs:
-            bc.apply(b)
-        from firedrake.assemble import _assemble
+    @cached_property
+    def _rhs(self):
+        from firedrake.assemble import create_assembly_callable
+        u = function.Function(self.trial_space)
+        b = function.Function(self.test_space)
         if isinstance(self.A.a, slate.TensorBase):
-            return _assemble(self.A.a * slate.AssembledVector(b))
+            expr = -self.A.a * slate.AssembledVector(u)
         else:
-            return _assemble(ufl.action(self.A.a, b))
+            expr = -ufl.action(self.A.a, u)
+        return u, create_assembly_callable(expr, tensor=b), b
+
+    def _lifted(self, b):
+        u, update, blift = self._rhs
+        u.dat.zero()
+        for bc in self.A.bcs:
+            bc.apply(u)
+        update()
+        # blift contains -A u_bc
+        blift += b
+        for bc in self.A.bcs:
+            bc.apply(blift)
+        # blift is now b - A u_bc, and satisfies the boundary conditions
+        return blift
```

Comment:
OK, so when boundary conditions are present, you populate `u` with the `bc` data and assemble the update into the rhs. The only real difference here is that you're always applying the bcs in case they have changed.

Reply:
OK, so if there are boundary conditions, then the modified right hand side is `b_orig - A u_gamma` where `u_gamma` is zero in the bulk and satisfies the boundary conditions on the boundary.

So we would do this modification before, but precompute `A u_gamma` once.

But that is unsafe, because what if the values in `u_gamma` change?  So the change here is that we compute `A u_gamma` every time we solve.